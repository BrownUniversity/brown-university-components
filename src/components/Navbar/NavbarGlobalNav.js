import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";
import { WindowSize } from "react-fns";

import NavbarContext from "./NavbarContext";
import Hamburger from "../Hamburger";
import Nav from "../Nav";
import Collapse from "../utils/Collapse";
import VisibilityToggle from "../utils/VisibilityToggle";
import colors from "../../constants/colors";
import media from "../../constants/media";
import { sansBold } from "../../constants/typography";

/*
  inner components
*/
const ToggleButton = styled.button`
  background-color: transparent;
  border: none;
  cursor: pointer;
  font-size: 1em;
  padding: 0;
`;

// filter props so they don't become dom attributes (see `styled-components` issue 439)
const ToggleTitle = styled(({ color, ...restProps }) => (
  <span {...restProps} />
))`
  color: ${({ color }) => colors[color]};
  font-family: ${sansBold};
  font-weight: bold;
  letter-spacing: 0.5px;
  margin-left: 10px;
  text-transform: uppercase;
`;

// filter props so they don't become dom attributes (see `styled-components` issue 439)
const CollapseWrapper = styled(({ color, ...restProps }) => (
  <div {...restProps} />
))`
  box-shadow: 0 5px 10px 0 #00000026;
  left: 0;
  position: absolute;
  top: 75px;
  width: 100%;
  background-color: ${({ color }) => colors[color]};
`;

const NavWrapper = styled.div`
  padding: 0 7vw 1rem;

  ${media.md`
    padding: 0 33px 1rem 33px;
  `};
`;

/*
  outer NavbarGlobalNav component
*/
const getChildColor = color => {
  switch (color) {
    case "white":
      return "red";

    // brown
    default:
      return "white";
  }
};

const NavbarGlobalNav = ({ children, ...restProps }) => (
  <WindowSize
    render={({ width }) => {
      // TODO: update when `width` doesn't return 0 on initial render (see `react-fns` issue 84)
      const currentWidth = width === 0 ? window.innerWidth : width;

      return (
        <NavbarContext.Consumer>
          {({ color, mobileBreakpoint, toggleTitle }) => {
            const childColor = getChildColor(color);
            const renderMobile = currentWidth < mobileBreakpoint;

            return (
              <div {...restProps}>
                <VisibilityToggle>
                  {({ isOpen: navIsOpen, toggleIsOpen: toggleNav }) => (
                    <React.Fragment>
                      <ToggleButton
                        type="button"
                        aria-controls="navbar-global-nav-collapse"
                        aria-expanded={navIsOpen}
                        aria-label="Toggle global navigation"
                        onClick={toggleNav}
                      >
                        <Hamburger
                          tag="div"
                          color={childColor}
                          isOpen={navIsOpen}
                        />
                        {!renderMobile && (
                          <ToggleTitle color={childColor}>
                            {toggleTitle}
                          </ToggleTitle>
                        )}
                      </ToggleButton>
                      <CollapseWrapper color={color}>
                        <Collapse
                          id="navbar-global-nav-collapse"
                          isOpen={navIsOpen}
                        >
                          <NavWrapper>
                            <Nav mobile color={childColor}>
                              {children}
                            </Nav>
                          </NavWrapper>
                        </Collapse>
                      </CollapseWrapper>
                    </React.Fragment>
                  )}
                </VisibilityToggle>
              </div>
            );
          }}
        </NavbarContext.Consumer>
      );
    }}
  />
);

NavbarGlobalNav.propTypes = {
  children: PropTypes.node.isRequired
};

NavbarGlobalNav.Item = Nav.Item;
NavbarGlobalNav.Link = Nav.Link;

export default NavbarGlobalNav;
