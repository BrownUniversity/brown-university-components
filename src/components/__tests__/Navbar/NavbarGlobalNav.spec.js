import React from "react";
import { breakpoints } from "brown-university-styles";
import { render, fireEvent } from "@testing-library/react";
import { triggerWindowResize, resetWindowSize } from "window-test-utils";
import NavbarContext from "../../Navbar/NavbarContext";
import NavbarGlobalNav from "../../Navbar/NavbarGlobalNav";

const renderNavbarGlobalNav = ({
  context = {},
  props = {},
  children = (
    <React.Fragment>
      <NavbarGlobalNav.Item>
        <NavbarGlobalNav.Link>Link</NavbarGlobalNav.Link>
      </NavbarGlobalNav.Item>
      <NavbarGlobalNav.Item>
        <NavbarGlobalNav.Link active>Active Link</NavbarGlobalNav.Link>
      </NavbarGlobalNav.Item>
      <NavbarGlobalNav.Item>
        <NavbarGlobalNav.Link disabled>Link</NavbarGlobalNav.Link>
      </NavbarGlobalNav.Item>
    </React.Fragment>
  ),
} = {}) => {
  const {
    color = "brown",
    mobileBreakpoint = breakpoints.md,
    toggleTitle = "Global Navigation",
  } = context;
  const rtlUtils = render(
    <NavbarContext.Provider value={{ color, mobileBreakpoint, toggleTitle }}>
      <NavbarGlobalNav {...props}>{children}</NavbarGlobalNav>
    </NavbarContext.Provider>,
  );
  const navbarGlobalNav = rtlUtils.container.firstChild;

  return {
    navbarGlobalNav,
    navbarGlobalNavToggleTitle: navbarGlobalNav.firstChild.children[1],
    navbarGlobalNavCollapseWrapper: navbarGlobalNav.children[1],
    ...rtlUtils,
  };
};

describe("NavbarGlobalNav", () => {
  describe("when window width is above mobileBreakpoint", () => {
    describe("styles", () => {
      it("should render brown navbar global nav by default", () => {
        const { navbarGlobalNav } = renderNavbarGlobalNav();

        expect(navbarGlobalNav).toMatchSnapshot();
      });

      it("should render white navbar global nav when color variant is provided", () => {
        const { navbarGlobalNavToggleTitle, navbarGlobalNavCollapseWrapper } =
          renderNavbarGlobalNav({
            context: { color: "white" },
          });

        expect(navbarGlobalNavToggleTitle).toHaveStyleRule("color", "#C00404");
        expect(navbarGlobalNavCollapseWrapper).toHaveStyleRule(
          "background-color",
          "#FFFFFF",
        );
      });
    });

    it("should render custom toggle title when provided", () => {
      const { getByText } = renderNavbarGlobalNav({
        context: { toggleTitle: "Custom Navigation" },
      });

      expect(getByText("Custom Navigation")).toBeInTheDocument();
    });

    it("should toggle navbar global nav collapse", () => {
      const { getByLabelText, getByText } = renderNavbarGlobalNav();
      const toggleButtonByLabelText = getByLabelText(
        "Toggle global navigation",
      );
      const toggleButtonByText = getByText("Global Navigation");

      expect(toggleButtonByLabelText).toHaveAttribute("aria-expanded", "false");
      fireEvent.click(toggleButtonByLabelText);
      expect(toggleButtonByLabelText).toHaveAttribute("aria-expanded", "true");
      fireEvent.click(toggleButtonByText);
      expect(toggleButtonByLabelText).toHaveAttribute("aria-expanded", "false");
    });
  });

  describe("when window width is below mobileBreakpoint", () => {
    beforeEach(() => {
      triggerWindowResize({ width: breakpoints.md - 1 });
    });

    afterAll(() => {
      resetWindowSize();
    });

    describe("styles", () => {
      it("should render brown mobile navbar global nav by default", () => {
        const { navbarGlobalNav } = renderNavbarGlobalNav();

        expect(navbarGlobalNav).toMatchSnapshot();
      });

      it("should render white mobile navbar global nav when color variant is provided", () => {
        const { navbarGlobalNavCollapseWrapper } = renderNavbarGlobalNav({
          context: { color: "white" },
        });

        expect(navbarGlobalNavCollapseWrapper).toHaveStyleRule(
          "background-color",
          "#FFFFFF",
        );
      });
    });

    it("should toggle navbar global nav collapse", () => {
      const { getByLabelText } = renderNavbarGlobalNav();
      const toggleButtonByLabelText = getByLabelText(
        "Toggle global navigation",
      );

      expect(toggleButtonByLabelText).toHaveAttribute("aria-expanded", "false");
      fireEvent.click(toggleButtonByLabelText);
      expect(toggleButtonByLabelText).toHaveAttribute("aria-expanded", "true");
      fireEvent.click(toggleButtonByLabelText);
      expect(toggleButtonByLabelText).toHaveAttribute("aria-expanded", "false");
    });
  });
});
