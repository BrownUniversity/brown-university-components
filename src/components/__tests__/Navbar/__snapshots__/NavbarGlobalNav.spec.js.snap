// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavbarGlobalNav when window width is above mobileBreakpoint styles should render brown navbar global nav by default 1`] = `
.c1 {
  background: transparent;
  border: none;
  cursor: inherit;
  display: inline-block;
  padding: 0 25px 7px 0;
}

.c2 {
  height: 3px;
  position: absolute;
  width: 25px;
  background-color: #FFFFFF;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease;
  transition-timing-function: ease;
  -webkit-transition-duration: .22s;
  transition-duration: .22s;
  -webkit-transition-timing-function: cubic-bezier(0.55,0.055,0.675,0.19);
  transition-timing-function: cubic-bezier(0.55,0.055,0.675,0.19);
}

.c2.c2:before {
  height: 3px;
  position: absolute;
  width: 25px;
  background-color: #FFFFFF;
  content: "";
  display: block;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease;
  transition-timing-function: ease;
  -webkit-transition-duration: .15s;
  transition-duration: .15s;
  -webkit-transition: top 0.1s 0.25s ease-in,opacity 0.1s ease-in;
  transition: top 0.1s 0.25s ease-in,opacity 0.1s ease-in;
  top: -7px;
}

.c2.c2:after {
  height: 3px;
  position: absolute;
  width: 25px;
  background-color: #FFFFFF;
  content: "";
  display: block;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease;
  transition-timing-function: ease;
  -webkit-transition-duration: .15s;
  transition-duration: .15s;
  -webkit-transition: bottom 0.1s 0.25s ease-in,-webkit-transform 0.22s cubic-bezier(0.55,0.055,0.675,0.19);
  -webkit-transition: bottom 0.1s 0.25s ease-in,transform 0.22s cubic-bezier(0.55,0.055,0.675,0.19);
  transition: bottom 0.1s 0.25s ease-in,transform 0.22s cubic-bezier(0.55,0.055,0.675,0.19);
  bottom: -7px;
}

.c8 {
  display: list-item;
  border-top: 0.5px solid #C8C8C880;
  padding: 1rem 0;
}

.c9 {
  background: transparent;
  border: none;
  display: block;
  font-family: CircularStd,Arial,Helvetica,sans-serif;
  padding: 0;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  color: #FFFFFF;
  cursor: pointer;
  font-size: 1em;
  font-weight: 700;
  line-height: 1.5;
  opacity: 1;
  pointer-events: auto;
  -webkit-transition: all .2s ease-in-out;
  transition: all .2s ease-in-out;
}

.c9:hover {
  color: #CCC6B8;
}

.c10 {
  background: transparent;
  border: none;
  display: block;
  font-family: CircularStd,Arial,Helvetica,sans-serif;
  padding: 0;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  color: #CCC6B8;
  cursor: pointer;
  font-size: 1em;
  font-weight: 700;
  line-height: 1.5;
  opacity: 1;
  pointer-events: auto;
  -webkit-transition: all .2s ease-in-out;
  transition: all .2s ease-in-out;
}

.c10:hover {
  color: #CCC6B8;
}

.c11 {
  background: transparent;
  border: none;
  display: block;
  font-family: CircularStd,Arial,Helvetica,sans-serif;
  padding: 0;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  color: #FFFFFF;
  cursor: not-allowed;
  font-size: 1em;
  font-weight: 700;
  line-height: 1.5;
  opacity: 1;
  pointer-events: auto;
  -webkit-transition: all .2s ease-in-out;
  transition: all .2s ease-in-out;
}

.c11:hover {
  color: #FFFFFF;
}

.c7 {
  list-style: none;
  margin: 0;
  background-color: transparent;
  display: block;
  padding: 0;
  width: 100%;
}

.c5 {
  will-change: height;
  display: none;
  height: 0;
  overflow: hidden;
  -webkit-transition: height 250ms cubic-bezier(0.4,0,0.2,1);
  transition: height 250ms cubic-bezier(0.4,0,0.2,1);
}

.c0 {
  background-color: transparent;
  border: none;
  cursor: pointer;
  font-size: 1em;
  padding: 0;
}

.c3 {
  color: #FFFFFF;
  font-family: CircularStdBold,CircularStd,Arial,Helvetica,sans-serif;
  font-weight: bold;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  margin-left: 10px;
  text-transform: uppercase;
}

.c4 {
  box-shadow: 0 5px 10px 0 #00000026;
  left: 0;
  position: absolute;
  top: 75px;
  width: 100%;
  background-color: #4E3629;
}

.c6 {
  padding: 0 7vw 1rem;
}

@media (min-width:740px) {
  .c6 {
    padding: 0 33px 1rem 33px;
  }
}

<div>
  <button
    aria-controls="navbar-global-nav-collapse"
    aria-expanded="false"
    aria-label="Toggle global navigation"
    class="c0"
    type="button"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      />
    </div>
    <span
      class="c3"
    >
      Global Navigation
    </span>
  </button>
  <div
    class="c4"
  >
    <div
      class="c5"
      display="none"
      height="0"
      id="navbar-global-nav-collapse"
      overflow="hidden"
    >
      <div
        class="c6"
      >
        <ul
          class="c7"
        >
          <li
            class="c8"
          >
            <button
              class="c9"
            >
              Link
            </button>
          </li>
          <li
            class="c8"
          >
            <button
              class="c10"
            >
              Active Link
            </button>
          </li>
          <li
            class="c8"
          >
            <button
              class="c11"
              disabled=""
            >
              Link
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`NavbarGlobalNav when window width is below mobileBreakpoint styles should render brown mobile navbar global nav by default 1`] = `
.c1 {
  background: transparent;
  border: none;
  cursor: inherit;
  display: inline-block;
  padding: 0 25px 7px 0;
}

.c2 {
  height: 3px;
  position: absolute;
  width: 25px;
  background-color: #FFFFFF;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease;
  transition-timing-function: ease;
  -webkit-transition-duration: .22s;
  transition-duration: .22s;
  -webkit-transition-timing-function: cubic-bezier(0.55,0.055,0.675,0.19);
  transition-timing-function: cubic-bezier(0.55,0.055,0.675,0.19);
}

.c2.c2:before {
  height: 3px;
  position: absolute;
  width: 25px;
  background-color: #FFFFFF;
  content: "";
  display: block;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease;
  transition-timing-function: ease;
  -webkit-transition-duration: .15s;
  transition-duration: .15s;
  -webkit-transition: top 0.1s 0.25s ease-in,opacity 0.1s ease-in;
  transition: top 0.1s 0.25s ease-in,opacity 0.1s ease-in;
  top: -7px;
}

.c2.c2:after {
  height: 3px;
  position: absolute;
  width: 25px;
  background-color: #FFFFFF;
  content: "";
  display: block;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease;
  transition-timing-function: ease;
  -webkit-transition-duration: .15s;
  transition-duration: .15s;
  -webkit-transition: bottom 0.1s 0.25s ease-in,-webkit-transform 0.22s cubic-bezier(0.55,0.055,0.675,0.19);
  -webkit-transition: bottom 0.1s 0.25s ease-in,transform 0.22s cubic-bezier(0.55,0.055,0.675,0.19);
  transition: bottom 0.1s 0.25s ease-in,transform 0.22s cubic-bezier(0.55,0.055,0.675,0.19);
  bottom: -7px;
}

.c7 {
  display: list-item;
  border-top: 0.5px solid #C8C8C880;
  padding: 1rem 0;
}

.c8 {
  background: transparent;
  border: none;
  display: block;
  font-family: CircularStd,Arial,Helvetica,sans-serif;
  padding: 0;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  color: #FFFFFF;
  cursor: pointer;
  font-size: 1em;
  font-weight: 700;
  line-height: 1.5;
  opacity: 1;
  pointer-events: auto;
  -webkit-transition: all .2s ease-in-out;
  transition: all .2s ease-in-out;
}

.c8:hover {
  color: #CCC6B8;
}

.c9 {
  background: transparent;
  border: none;
  display: block;
  font-family: CircularStd,Arial,Helvetica,sans-serif;
  padding: 0;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  color: #CCC6B8;
  cursor: pointer;
  font-size: 1em;
  font-weight: 700;
  line-height: 1.5;
  opacity: 1;
  pointer-events: auto;
  -webkit-transition: all .2s ease-in-out;
  transition: all .2s ease-in-out;
}

.c9:hover {
  color: #CCC6B8;
}

.c10 {
  background: transparent;
  border: none;
  display: block;
  font-family: CircularStd,Arial,Helvetica,sans-serif;
  padding: 0;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  color: #FFFFFF;
  cursor: not-allowed;
  font-size: 1em;
  font-weight: 700;
  line-height: 1.5;
  opacity: 1;
  pointer-events: auto;
  -webkit-transition: all .2s ease-in-out;
  transition: all .2s ease-in-out;
}

.c10:hover {
  color: #FFFFFF;
}

.c6 {
  list-style: none;
  margin: 0;
  background-color: transparent;
  display: block;
  padding: 0;
  width: 100%;
}

.c4 {
  will-change: height;
  display: none;
  height: 0;
  overflow: hidden;
  -webkit-transition: height 250ms cubic-bezier(0.4,0,0.2,1);
  transition: height 250ms cubic-bezier(0.4,0,0.2,1);
}

.c0 {
  background-color: transparent;
  border: none;
  cursor: pointer;
  font-size: 1em;
  padding: 0;
}

.c3 {
  box-shadow: 0 5px 10px 0 #00000026;
  left: 0;
  position: absolute;
  top: 75px;
  width: 100%;
  background-color: #4E3629;
}

.c5 {
  padding: 0 7vw 1rem;
}

@media (min-width:740px) {
  .c5 {
    padding: 0 33px 1rem 33px;
  }
}

<div>
  <button
    aria-controls="navbar-global-nav-collapse"
    aria-expanded="false"
    aria-label="Toggle global navigation"
    class="c0"
    type="button"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      />
    </div>
  </button>
  <div
    class="c3"
  >
    <div
      class="c4"
      display="none"
      height="0"
      id="navbar-global-nav-collapse"
      overflow="hidden"
    >
      <div
        class="c5"
      >
        <ul
          class="c6"
        >
          <li
            class="c7"
          >
            <button
              class="c8"
            >
              Link
            </button>
          </li>
          <li
            class="c7"
          >
            <button
              class="c9"
            >
              Active Link
            </button>
          </li>
          <li
            class="c7"
          >
            <button
              class="c10"
              disabled=""
            >
              Link
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
